src/Disks/DiskObjectStorage/DiskObjectStorageTransaction.cpp

```diff
@@ -897,6 +897,14 @@ std::unique_ptr<WriteBufferFromFileBase> DiskObjectStorageTransaction::writeFile
     if (mode == WriteMode::Append && !metadata_transaction->supportAddingBlobToMetadata())
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Disk does not support WriteMode::Append");
 
+    if (metadata_storage.isPlain())
+    {
+        /// Materialize virtual parent.
+        const auto parent_path = fs::path(path).parent_path();
+        if (metadata_storage.existsDirectory(parent_path))
+            createDirectories(parent_path);
+    }
```

Comment:
This should not be in `DiskObjectStorage` because it is a metadata storage responsibility for maintaining its layout correctly. 

Let's move this check into the `generateObjectKeyForPath`?